pkg/build/strategies/sti/postexecutorstep.go

```diff
@@ -0,0 +1,467 @@
+	fs      util.FileSystem
+}
+
+func (step *startRuntimeImageAndUploadFilesStep) execute(ctx *postExecutorStepContext) error {
+	glog.V(1).Info("Executing step: start runtime image and upload files")
+
+	fd, err := ioutil.TempFile("", "s2i-upload-done")
+	if err != nil {
+		return err
+	}
+	fd.Close()
+	lastFilePath := fd.Name()
+	defer func() {
+		os.Remove(lastFilePath)
+	}()
+
+	lastFileDstPath := "/tmp/" + filepath.Base(lastFilePath)
+
+	outReader, outWriter := io.Pipe()
+	defer outReader.Close()
+	defer outWriter.Close()
+
+	errReader, errWriter := io.Pipe()
+	defer errReader.Close()
+	defer errWriter.Close()
+
+	artifactsDir := filepath.Join(step.builder.config.WorkingDir, api.RuntimeArtifactsDir)
+
+	scripts := []string{api.Run}
+	useExternalAssembleScript := step.builder.externalScripts[api.AssembleRuntime]
+	if useExternalAssembleScript {
+		scripts = append(scripts, api.AssembleRuntime)
+	}
+
+	// We copy scripts to a directory with artifacts to upload files in one shot
+	for _, script := range scripts {
+		src := filepath.Join(step.builder.config.WorkingDir, api.UploadScripts, script)
+		dst := filepath.Join(artifactsDir, script)
+		glog.V(2).Infof("Copying file %q -> %q", src, dst)
+		if err := step.fs.Copy(src, dst); err != nil {
+			return fmt.Errorf("Couldn't copy file (%q -> %q): %v", src, dst, err)
+		}
+	}
+
+	image := step.builder.config.RuntimeImage
+	workDir, err := step.docker.GetImageWorkdir(image)
+	if err != nil {
+		return fmt.Errorf("Couldn't get working dir of %q image: %v", image, err)
+	}
+
+	commandBaseDir := workDir
+	if !useExternalAssembleScript {
+		// script already inside of the image
+		scriptsURL, err := step.docker.GetScriptsURL(image)
+		if err != nil {
+			return err
+		}
+		if len(scriptsURL) == 0 {
+			return fmt.Errorf("Couldn't determine scripts URL for image %q", image)
+		}
+		commandBaseDir = strings.TrimPrefix(scriptsURL, "image://")
+	}
+
+	waitForFile := fmt.Sprintf("while [ ! -f %q ]; do sleep 0.5; done", lastFileDstPath)
+	runScript := fmt.Sprintf("%s/%s", commandBaseDir, api.AssembleRuntime)
+	rememberExitCode := "result=$?"
+	cleanup := fmt.Sprintf("rm -f -- %q", lastFileDstPath)
+	if useExternalAssembleScript {
+		cleanup += fmt.Sprintf(" %q", runScript)
+	}
+	exit := "exit $result"
+	cmd := strings.Join([]string{waitForFile, runScript, rememberExitCode, cleanup, exit}, "; ")
+
+	opts := dockerpkg.RunContainerOptions{
+		Image:           image,
+		PullImage:       false, // The PullImage is false because we've already pulled the image
+		CommandExplicit: []string{"/bin/sh", "-c", cmd},
+		Stdout:          outWriter,
+		Stderr:          errWriter,
+		NetworkMode:     string(step.builder.config.DockerNetworkMode),
+		CGroupLimits:    step.builder.config.CGroupLimits,
+		CapDrop:         step.builder.config.DropCapabilities,
+		PostExec:        step.builder.postExecutor,
+		Env:             step.builder.env,
+		User:            "root", // without this we won't be able to remove last file
```

Comment:
@bparees under what user we should run runtime image with `assemble-runtime` script? Currently it runs under `root` which also allows us to remove temporary file (that is used as signal that uploading done).